### ISSUE TITLE
Can procedure-reduce-arity preserve equality sometimes?

### ISSUE THREAD

**Issue by jackfirth** [racket/racket#3042]
In [Rebellion](https://github.com/jackfirth/rebellion), I occasionally need to write code like this:

```racket
(if (equal? (procedure-arity f) 2)
    f
    (procedure-reduce-arity f 2))
```

This is usually because:

1. I'm stuffing the function into a transparent struct
2. All functions in that struct should have a specific arity
3. I want to write chaperone contracts on those structs that wrap the function inside with a chaperone while still preserving `equal?`.

The chaperone contracts wrap the functions with guards on their inputs/outputs, and they assume the functions always have the arity they're supposed to. But if I _always_ call `procedure-reduce-arity`, then the resulting wrapper function is not `equal?` to the input even if it's arity is unchanged. This makes chaperone contracts fail.

You can see an example of this problem in the implementation of [comparators](https://github.com/jackfirth/rebellion/blob/master/private/comparator.rkt) in Rebellion—simply replace the `(if (equal? (procedure-arity ...)))` expression with an unconditional call to `procedure-reduce-arity`. That should cause failures related to `make-chaperone-contract` and `comparator/c`.

Can the contract of `procedure-reduce-arity` be changed to promise that the returned procedure will be `equal?` to the input procedure if the input procedure already has the desired arity?

**Comment by mflatt** [comment]
An obstacle to that change is that `procedure-reduce-arity` currently can be used to hide the structness of an applicable struct. I'm not sure I've ever used it that way, but I think I might have. So, the change would be backward-incompatible.

**Comment by jackfirth** [comment]
Are there other, better ways to do that? Or is using `procedure-reduce-arity` the best way to do that today?

**Comment by NoahStoryM** [comment]
Hi @mflatt , I have a similar situation and I wonder if we can add `#:new?` to `procedure-reduce-arity`, `procedure-rename`, etc to solve this problem?

**Comment by mflatt** [comment]
Thinking about this more, the reason `procedure-reduce-arity` exists is so it affects the result of functions like `procedure-arity`...

Otherwise, you can change the number of arguments that a procedure is willing to accept by wrapping it with a `case-lambda` form that passes a "good" number of arguments to the original procedure and raises an exception for a "bad" number of arguments. And you can preserve `equal?` for that wrapper by creating it with `chaperone-procedure` (using a procedure created with `case-lambda` as the second argument to `chaperone-procedure`). Either way, `procedure-arity` will report the wider interface for the wrapped procedures — and `chaperone-procedure` will ensure that exactly the original procedure's arity is reported, even though some argument counts will trigger an exception.

Part of the justification for preserving `equal?` via chaperones is that they change behavior only by raising additional exceptions, and that's a carefully chosen generalization of "equal". Allowing `procedure-reduce-arity` (or `chaperone-procedure`) to also change the result of `procedure-arity` seems like a big additional change. Maybe its justified on the grounds that `procedure-arity` exists to report errors earlier, but I'm not sure.

cc @rfindler

**Comment by rfindler** [comment]
I agree with everything @mflatt writes in the comment just above but I have always wondered why `procedure-reduce-arity` (and other procedure-related functions) hide some aspects of procedures while also doing their jobs (in this case, reducing the arity). I believe there is even a way to get the contract library in trouble along these lines (but I don't think it is because the contract library calls `procedure-reduce-arity` specifically; I cannot recall what the issue was exactly any more).

That said, it might make sense to have a shortcircuit in procedure-reduce-arity so that if the arity is already the same and if the name is the same and if the realm is the same then it returns an equal? procedure. I'm not sure this is worth it, however, since the default values for realm and name aren't the current name and realm of the procedure (so it won't help too much on the client side). It is like the optimization of checking to see if the second argument to `append` (in a two argument call) is '() and returning a list eq? to the first argument. Yes, it is helpful in some situations but it might also surprise someone down the line.

**Comment by jackfirth** [comment]
> Part of the justification for preserving `equal?` via chaperones is that they change behavior only by raising additional exceptions, and that's a carefully chosen generalization of "equal". Allowing `procedure-reduce-arity` (or `chaperone-procedure`) to also change the result of `procedure-arity` seems like a big additional change. Maybe its justified on the grounds that `procedure-arity` exists to report errors earlier, but I'm not sure.

This doesn't matter for cases where `procedure-reduce-arity` isn't changing the function's arity, correct? That's clearly a behavior-preserving change - it doesn't affect the function's runtime behavior for any inputs and it doesn't change the result of `(procedure-arity f)`. I assume that `(procedure-reduce-arity (procedure-reduce-arity f x) x)` behaves identically to `(procedure-reduce-arity f x)` in all ways _except_ that they're not `equal?`. They both report a `procedure-arity` of `x`, and they both throw the same errors when given other numbers of arguments.

**Comment by mflatt** [comment]
> This doesn't matter for cases where `procedure-reduce-arity` isn't changing the function's arity, correct?

I had indeed lost track of the original point. A `new?` argument where `#f` matters only when the arity (and name and realm) stay the same would be fine and fully compatible.
